### ISSUE TITLE
Implement a probability calibration scheme

### ISSUE THREAD

**Issue by karlnapf** [shogun-toolbox/shogun#3851]
Shogun needs a way to turn calibrated probability outputs of its models (SVM, RF, etc), into calibrated ones. This means that the predicted class probability actually matches the actual class probability (as estimated over some validation dataset)

Inspiration:
http://scikit-learn.org/stable/modules/calibration.html

This requires some form of x-validation, so can be tied into Shogun's framework for that

Somehow related to #3850

**Comment by karlnapf** [comment]
@durovo That would be a great addition to Shogun, such a welcome contribution!

It is a bit more tricky, make sure to share early drafts on this soon. Don't get into writing code before we have discussed a design draft. I suggest to read our corss-validation code before you start.

**Comment by durovo** [comment]
@karlnapf Understood. I am currently going through the cross-validation code. I will share a rough draft as soon as I have figured it out. Should I use this thread to discuss the design?

**Comment by karlnapf** [comment]
Yep this thread is totally fine. I would also be interested in hearing about the sklearn approach to see whether we can learn from it. I guess it would be best to re-use the x-validation code.

**Comment by durovo** [comment]
@karlnapf I have gone through the sklearn approach and the shogun code related to this problem. Sklearn is using two methods for probability calibration:
1. Pratts scaling 
2. Isotonic regression
The `BinaryLabels` class in shogun already does probability calibration in the function `scores_to_probabilities`,  which uses `CStatistics::fit_sigmoid` for the same. 

The sklearn approach:

1. CalibratorCV has the same interface as that of estimators.
2. An estimator and a cross-validation strategy must be passed to the CalibratorCV.
3. The x-validation strategy is later used to split the dataset and labels when `fit` is called on them
4. For each fold, the estimator is trained on the training data. One of the two approaches given above is used to fit the output of the estimator on test data to the test labels. 
5. The calibrator object stores a copy of the trained estimator, this calibrator object is appended to a global list.
6. Later, when predict_probability is called on some data: 
  For each calibrator in the list:
    1. The estimator is used to predict the probabilities.
    2. These are then fed to the fitted calibration method to generate scaled probabilities.
7. The mean over all estimator-calibrator pairs is then returned.

**Comment by karlnapf** [comment]
Any news on this?

**Comment by durovo** [comment]
@karlnapf Sadly, I have been unable to work on this due to ongoing exams. I was trying to use the existing cross-validation code inside a class that inherits from the Machine class to make an interface similar to the one in scikit-learn. Can you point me to some resource/documentation on the locks that are being used everywhere in shogun? I am having a hard time understanding the logic behind them. Are they being used to ensure that parallel execution doesn't cause any problems?

**Comment by karlnapf** [comment]
No worries about the delay, your work and thoughts are still very much appreciated!
I think your thinking makes sense, maybe we move on towards designing some header/API files?

Don't worry about the parallel thing for now, we can address that once the design is there.

**Comment by durovo** [comment]
Hello @karlnapf , I'm almost done with the implementation part I believe. Just need to fix a few things and write tests before making the PR. 
About the API, I have tried to maintain the same interface as that of CMachine. I have named the class CCrossValidatedCalibration. The user can call train or train_machine to the run the cross-validation in which for each subset a new Calibrator will be generated. (A calibrator basically implements the calibration part in its train function, I have written a separate class for it. We can then allow the user to pass a Calibrator of their choice along with their machine, features, and labels to CrossValidatedCalibration). 
The apply and apply_locked will return the mean of calibrated labels predicted by the Calibrator machines.

Should I just post the code here so that you'll be able to see the intended design?

**Comment by durovo** [comment]
Any suggestions on how to write the tests for this?

**Comment by karlnapf** [comment]
The first PR is just for discussion and we can first discuss the API before you write tests.

Having said that, minimal tests are essential, try to cover things like: preserving order, no leak of future data. Best thing is to do a manual split by pen&paper and then assert things in the unit test. Next thing is to think about corner cases ...

**Comment by durovo** [comment]
@karlnapf I have created a pull request so that we can discuss the API.
